ifrs17/Report/ReportMutableScopesInteractive.ipynb

```diff
@@ -0,0 +1,529 @@
+            "source": [
+                "[InitializeScope(nameof(InitReportStorageScopeAsync))]",
+                "\npublic interface ReportScope : IMutableScope<string>, ReportingNodeFormsEntity, MonthlyPeriodFormsEntity, ScenarioFormsEntity, CurrencyFormsEntity, BasicSliceAndDiceFormsEntity, BasicFilterFormsEntity {",
+                "\n    protected Systemorph.Vertex.Pivot.Builder.Interfaces.IPivotFactory report => GetStorage().Report;",
+                "\n    protected int headerColumnWidthValue => 250;",
+                "\n",
+                "\n    HashSet<(ReportIdentity, CurrencyType)> GetDataIdentities() => GetStorage().GetIdentities((Year, Month), ReportingNode, Scenario, CurrencyType); // TODO, add filter for identities, if the property is exposed at this level",
+                "\n       ",
+                "\n    async Task<GridOptions> GetReportTaskAsync(IDataCube<ReportVariable> data) {",
+                "\n        return await report.ForDataCube(data)",
+                "\n            .WithQuerySource(workspace)",
+                "\n            .SliceRowsBy(rowSlices)",
+                "\n            .SliceColumnsBy(columnSlices)",
+                "\n            .ReportGridOptions(headerColumnWidth: headerColumnWidthValue)",
+                "\n            .ExecuteAsync();",
+                "\n    }",
+                "\n",
+                "\n    async Task InitReportStorageScopeAsync() { // This has the Async issue, but imo it should come in the future",
+                "\n       await GetStorage().InitializeReportIndependentCacheAsync();",
+                "\n    }",
+                "\n}"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public interface Data : IMutableScope<((int year, int month) reportingPeriod, string reportingNode, string scenario, CurrencyType currencyType, (string filterName, object filterValue)[] dataFilter)> {",
+                "\n    IDataCube<ReportVariable> InputDataCube { get; set; }",
+                "\n",
+                "\n    IDataCube<ReportVariable> DataCube { get {",
+                "\n        if(InputDataCube is null) return Enumerable.Empty<ReportVariable>().ToDataCube();",
+                "\n        var filteredDataCube = (Identity.dataFilter is null || Identity.dataFilter.Length == 0) ? InputDataCube : InputDataCube.Filter(Identity.dataFilter); ",
+                "\n        if(Identity.scenario != \"Delta\") return filteredDataCube;",
+                "\n        var bestEstimateById = filteredDataCube.Where(x => x.Scenario == null).ToDictionary(x => x.ToIdentityString());",
+                "\n        return filteredDataCube.Select(x => x.Scenario == null ? x : x with { Value = x.Value - (bestEstimateById.TryGetValue((x with {Scenario = null}).ToIdentityString(), out var be)? be.Value : 0d) }).ToDataCube();",
```

Comment:
why not having the difference of two dataCubes? @amuolo - the identity properties should take care of doing the diff correctly.

Reply:
The diff between the 2 datacubes wouldn't return the desired data because we want to subtract each entry in the scenario data cube with the corresponding entry in the best estimate datacube. In fact, since the scenario dimension is different, by simply performing the difference between the 2 datacubes one wouldn't achieve the same result.